Add Dockerfile for independent parflow image

psavery opened the pull request:
This parflow image is intended to be ran as a separate image, rather
than being a part of our cluster image. It is based largely on [this Dockerfile](https://github.com/Kitware/hpccloud-services/blob/master/docker/compute-parflow/Dockerfile).

The runtime is created independently from the build image to
minimize the size of the final docker image, which is currently only
158 MB.

Some tests can be ran in the parflow repo's test directory, and they seem
to work. For instance:
```
docker run --rm -w /data -v $PWD:/data kitware/hpccloud:parflow default_single.tcl 1 1 1
```

This Dockerfile can be used to create other types of images as well, such
as singularity and shifter.

psavery committed:
Add Dockerfile for independent parflow image

This parflow image is intended to be ran as a separate image, rather
than being a part of our cluster image.

The runtime is created independently from the build image to
minimize the size of the final docker image, which is currently only
158 MB.

Some tests can be ran in the parflow repo's test directory, and they seem
to work. For instance:
```
docker run --rm -w /data -v $PWD:/data kitware/hpccloud:parflow default_single.tcl 1 1 1
```

This Dockerfile can be used to create other types of images as well, such
as singularity and shifter.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery commented:
To do:
- [x] Base the image off of Ubuntu 20.04 instead of 18.04
- [x] Run the container with a user rather than as root, to skip the openmpi root issues

psavery committed:
Use Ubuntu 20.04 as base for Parflow image

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
Run parflow as a non-root user

This allows us to avoid the openmpi options and environment variables
to allow it to run as root.

It adds a /scratch directory where the user may mount the input files.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

jourdain commented on docker/compute/parflow/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+      version="1.0" \
+      parflow.version="master"
+
+COPY --from=build /parflow /parflow
```
Don't you need anything else from the build? Such as `/hypre`?
Garrett may have some input too.

psavery replied:
So, I took a look inside `/hypre`, and it only appears to contain header files and a static library. I don't think it has any runtime dependencies like shared libraries, so I don't think it's needed at runtime.

I can run this image on their test examples, by the way, and it works for the ones I tried. See [here](https://github.com/Kitware/hpccloud-services/pull/14/files#diff-edc4fb9ca653e11c18b1f715d524018cR13) for an example.

jourdain commented on docker/compute/parflow/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+ENV PARFLOW_MPIEXEC_EXTRA_FLAGS "--mca mpi_yield_when_idle 1 --oversubscribe"
+
+# Run as a non-root user
+RUN useradd -m -U -s /bin/bash user
```
do we want the user to be `user` or `parflow` or `ubuntu` like on aws?

psavery replied:
It can be anything. Just let me know which one.

jourdain replied:
I agree with anything, but I was just wondering if there was a convention of some kind somewhere? Any previous experience on that?

Should we use a uid of 1000 too?

psavery replied:
I am not sure about a convention for names. I haven't personally seen a common convention.

I think making a uid of 1000 is probably a good idea, so we get a consistent UID. It is also what [jupyter does](https://github.com/jupyter/docker-stacks/blob/76402a27fd13e649a3166b1da5a162ac476912bf/base-notebook/Dockerfile#L13-L14), and I think lots of people use their docker images. I guess it might be a good convention for us to use a UID of 1000 and a GID of 100.

Note that they also use "jovyan" for their user names, which is just what jupyter uses to describe [members of the jupyter community](https://jupyter.readthedocs.io/en/latest/community/content-community.html#what-is-a-jovyan).

jourdain commented:
@grapp1

jourdain commented:
Then let’s go with user ubuntu like aws do for its ubuntu based images.

On Tue, Jun 2, 2020 at 22:29 Patrick Avery <notifications@github.com> wrote:

> *@psavery* commented on this pull request.
> ------------------------------
>
> In docker/compute/parflow/Dockerfile
> <https://github.com/Kitware/hpccloud-services/pull/14#discussion_r434157047>
> :
>
> > +ENV PARFLOW_DIR=/parflow
> +
> +# Install required runtime files
> +RUN apt-get update && \
> +    DEBIAN_FRONTEND=noninteractive \
> +    apt-get install -y --no-install-recommends \
> +        openmpi-bin \
> +        libgfortran-9-dev \
> +        openssh-server \
> +        tcl && \
> +    rm -rf /var/lib/apt/lists/*
> +
> +ENV PARFLOW_MPIEXEC_EXTRA_FLAGS "--mca mpi_yield_when_idle 1 --oversubscribe"
> +
> +# Run as a non-root user
> +RUN useradd -m -U -s /bin/bash user
>
> I am not sure about a convention for names. I haven't personally seen a
> common convention.
>
> I think making a uid of 1000 is probably a good idea, so we get a
> consistent UID. It is also what jupyter does
> <https://github.com/jupyter/docker-stacks/blob/76402a27fd13e649a3166b1da5a162ac476912bf/base-notebook/Dockerfile#L13-L14>,
> and I think lots of people use their docker images. I guess it might be a
> good convention for us to use a UID of 1000 and a GID of 100.
>
> Note that they also use "jovyan" for their user names, which is just what
> jupyter uses to describe members of the jupyter community
> <https://discourse.jupyter.org/t/what-is-with-the-weird-jovyan-user/1673/3>
> .
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Kitware/hpccloud-services/pull/14#discussion_r434157047>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACH45VPAIBGMNMU5UVCNVLRUVOL5ANCNFSM4NMD4OWA>
> .
>

psavery committed:
Use the user "ubuntu" instead of "user"

This is a convention that AWS uses.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery committed:
Use fixed uid and gid for docker user

We are using 1000 for the uid as jupyter uses it.

Jupyter uses 100 for the gid, but we are using 1000, because users
on computers typically have matching gid's, and I am not sure why
jupyter prefers 100 for the gid.

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery commented:
Done. I set the uid to be 1000 like jupyter does. But I set the gid to be 1000 instead of 100. I am not sure why jupyter uses a gid of 100. On most computers, it seems that the uid and gid of a user typically match. So I went with gid of 1000.

psavery commented:
Should we remove [this one](https://github.com/Kitware/hpccloud-services/blob/master/docker/compute-parflow/Dockerfile) as part of this PR as well? The old one uses the sge-ssh as a base image, whereas this one is stand-alone.

patrickoleary commented:
Yes, clean it up.

On Tue, Jun 2, 2020 at 6:53 PM Patrick Avery <notifications@github.com>
wrote:

> Should we remove this one
> <https://github.com/Kitware/hpccloud-services/blob/master/docker/compute-parflow/Dockerfile>
> as part of this PR as well? The old one uses the sge-ssh as a base image,
> whereas this one is stand-alone.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kitware/hpccloud-services/pull/14#issuecomment-637849013>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYGDMKBLLR77NLWIUGUBZLRUV7HFANCNFSM4NMD4OWA>
> .
>

jourdain commented:
> Most Linux distributions reserve the first 100 UIDs for system use. New users are assigned UIDs starting from 500 or 1000. For example, new users in Ubuntu start from 1000.

So 1000 seems reasonable as it will be the very first user which will implicitly match the main user of the host system.

psavery committed:
Merge branch 'master' into parflow-docker

psavery committed:
Remove old compute-parflow and clarify new style

Signed-off-by: Patrick Avery <patrick.avery@kitware.com>

psavery commented:
So, just to record this here, I think our goal is to eventually have a single slurm cluster running in hpccloud-services (right now, there are two: compute and visualize), and then be able to run (using docker or singularity) on the slurm cluster images like this one.

So, the slurm cluster will come with no simulation/visualization software installed on it, and running simulations or visualizations would consist of running docker images (like this parflow one).

jourdain commented:
I guess my assumption was to have code bundled inside singularity images rather than docker, but in general, yes to what you just said. A generic cluster on which you can run things on it using some container encapsulation.

psavery commented:
Yep, singularity will work too!
